dbms/src/Storages/Page/V3/PageDirectory.cpp

```diff
@@ -1240,78 +1240,86 @@ void PageDirectory::applyRefEditRecord(
+                switch (r.type)
+                {
+                case EditRecordType::PUT_EXTERNAL:
                 {
-                    // put the new created holder into `external_ids`
-                    *holder = r.page_id;
-                    external_ids_by_ns.addExternalId(holder);
+                    auto holder = version_list->createNewExternal(r.version);
+                    if (holder)
+                    {
+                        // put the new created holder into `external_ids`
+                        *holder = r.page_id;
+                        external_ids_by_ns.addExternalId(holder);
+                    }
+                    break;
+                }
+                case EditRecordType::PUT:
+                    version_list->createNewEntry(r.version, r.entry);
+                    break;
+                case EditRecordType::DEL:
+                    version_list->createDelete(r.version);
+                    break;
+                case EditRecordType::REF:
+                    applyRefEditRecord(mvcc_table_directory, version_list, r, r.version);
+                    break;
+                case EditRecordType::UPSERT:
+                case EditRecordType::VAR_DELETE:
+                case EditRecordType::VAR_ENTRY:
+                case EditRecordType::VAR_EXTERNAL:
+                case EditRecordType::VAR_REF:
+                    throw Exception(fmt::format("should not handle edit with invalid type [type={}]", magic_enum::enum_name(r.type)));
                 }
-                break;
             }
-            case EditRecordType::PUT:
-                version_list->createNewEntry(r.version, r.entry);
-                break;
-            case EditRecordType::DEL:
-                version_list->createDelete(r.version);
-                break;
-            case EditRecordType::REF:
-                applyRefEditRecord(mvcc_table_directory, version_list, r, r.version);
-                break;
-            case EditRecordType::UPSERT:
-            case EditRecordType::VAR_DELETE:
-            case EditRecordType::VAR_ENTRY:
-            case EditRecordType::VAR_EXTERNAL:
-            case EditRecordType::VAR_REF:
-                throw Exception(fmt::format("should not handle edit with invalid type [type={}]", magic_enum::enum_name(r.type)));
+            catch (DB::Exception & e)
+            {
+                e.addMessage(fmt::format(" [type={}] [page_id={}] [ver={}] [edit_size={}]", magic_enum::enum_name(r.type), r.page_id, r.version, edit_size));
+                e.rethrow();
             }
         }
-        catch (DB::Exception & e)
-        {
-            e.addMessage(fmt::format(" [type={}] [page_id={}] [ver={}] [edit_size={}]", magic_enum::enum_name(r.type), r.page_id, r.version, edit.size()));
-            e.rethrow();
-        }
-    }
 
-    // stage 3, the edit committed, incr the sequence number to publish changes for `createSnapshot`
-    sequence.fetch_add(new_sequence - last_sequence);
```

Comment:
This will make `sequence` increase by `edit_size` + 1. Please check if it is necessary to add the 1. @JaySon-Huang

Reply:
After some thought, we do not need to add `sequence` for an empty WriteBatch. So it is not necessary, increase by edit_size is OK.